File: benchmarks/python/bench_ctoon.py
#!/usr/bin/env python3
"""CToon Python benchmark — vs. competing implementations.

Same methodology as every other language benchmark in this repo (see
benchmarks/README.md), so the numbers line up with the C, C++, and Go
results for the same corpus:

  1. Load every file in the corpus manifest into memory (untimed).
  2. Untimed pre-pass: convert each JSON file to TOON once (with ctoon).
  3. Timed "JSON -> TOON": repeatedly parse JSON and re-serialise to TOON.
  4. Timed "TOON -> JSON": repeatedly parse the TOON text from step 2 and
     re-serialise to JSON.
  5. Report throughput (MB/s of the bytes actually read by that operation)
     and documents/sec.

Runs against every implementation importable in the environment:
  - ctoon              (this repo)
  - toon_format        github.com/toon-format/toon-python (official, pure Python)
  - toons              github.com/alesanfra/toons (community, Rust backend)
"""
import argparse
import json
import sys
import time

import ctoon

try:
    from tabulate import tabulate
except ImportError:
    tabulate = None

try:
    import toon_format
except ImportError:
    toon_format = None

try:
    import toons
except ImportError:
    toons = None

REPEATS = 20


def load_corpus(manifest_path):
    files = []
    with open(manifest_path, "r") as mf:
        paths = [line.strip() for line in mf if line.strip()]
    for path in paths:
        try:
            with open(path, "r", encoding="utf-8") as f:
                files.append({"path": path, "json": f.read(), "toon": None})
        except OSError:
            continue
    return files


def bench_json_to_toon(files, json_to_toon_fn):
    """json_to_toon_fn(json_text) -> toon string."""
    t0 = time.perf_counter()
    ops = 0
    bytes_done = 0
    for _ in range(REPEATS):
        for f in files:
            try:
                json_to_toon_fn(f["json"])
                ops += 1
                bytes_done += len(f["json"].encode("utf-8"))
            except Exception:
                pass
    return time.perf_counter() - t0, ops, bytes_done


def bench_toon_to_json(files, toon_to_json_fn):
    """toon_to_json_fn(toon_text) -> json string."""
    t0 = time.perf_counter()
    ops = 0
    bytes_done = 0
    for _ in range(REPEATS):
        for f in files:
            if not f["toon"]:
                continue
            try:
                toon_to_json_fn(f["toon"])
                ops += 1
                bytes_done += len(f["toon"].encode("utf-8"))
            except Exception:
                pass
    return time.perf_counter() - t0, ops, bytes_done


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("manifest")
    args = parser.parse_args()

    files = load_corpus(args.manifest)
    if not files:
        print("Corpus manifest is empty - nothing to benchmark.", file=sys.stderr)
        return 1

    total_json_bytes = sum(len(f["json"].encode("utf-8")) for f in files)

    print("CToon Python Benchmark — vs. competing implementations")
    print(f"Corpus: {len(files)} files, {total_json_bytes / 1e6:.2f} MB (JSON)")
    libs = ["ctoon (this repo)"]
    if toon_format:
        libs.append("toon_format (official, pure Python)")
    if toons:
        libs.append("toons (community, Rust backend)")
    print("Implementations: " + ", ".join(libs) + "\n")

    # ── Untimed pre-pass: TOON text with ctoon, shared as decode input ──
    total_toon_bytes = 0
    for f in files:
        try:
            data = json.loads(f["json"])
            f["toon"] = ctoon.dumps(data)
            total_toon_bytes += len(f["toon"].encode("utf-8"))
        except Exception:
            pass

    rows = []

    def add_rows(name, json_to_toon_fn, toon_to_json_fn):
        t_enc, ops_enc, bytes_enc = bench_json_to_toon(files, json_to_toon_fn)
        rate_enc = f"{100 * ops_enc / (len(files) * REPEATS):.0f}%"
        rows.append([
            name, "JSON -> TOON",
            f"{bytes_enc / t_enc / 1e6:.2f} MB/s" if ops_enc else "n/a",
            f"{ops_enc / t_enc:.0f}",
            rate_enc,
            f"{t_enc:.4f} s",
        ])

        t_dec, ops_dec, bytes_dec = bench_toon_to_json(files, toon_to_json_fn)
        rate_dec = f"{100 * ops_dec / (len(files) * REPEATS):.0f}%"
        rows.append([
            name, "TOON -> JSON",
            f"{bytes_dec / t_dec / 1e6:.2f} MB/s" if ops_dec else "n/a",
            f"{ops_dec / t_dec:.0f}",
            rate_dec,
            f"{t_dec:.4f} s",
        ])

    add_rows(
        "ctoon",
        lambda text: ctoon.dumps(json.loads(text)),
        lambda text: ctoon.dumps_json(ctoon.loads(text), indent=2),
    )

    if toon_format:
        add_rows(
            "toon_format",
            lambda text: toon_format.encode(json.loads(text)),
            lambda text: json.dumps(toon_format.decode(text), indent=2),
        )

    if toons:
        add_rows(
            "toons",
            lambda text: toons.dumps(json.loads(text)),
            lambda text: toons.to_json(text, indent=2),
        )

    headers = ["Library", "Operation", "Throughput", "Docs/sec", "Success", f"Total time (x{REPEATS} reps)"]
    if tabulate:
        print(tabulate(rows, headers=headers))
    else:
        fmt = "{:<14} {:<14} {:>12} {:>10} {:>8} {:>24}"
        print(fmt.format(*headers))
        for row in rows:
            print(fmt.format(*row))

    return 0


if __name__ == "__main__":
    sys.exit(main())

File: benchmarks/cpp/bench_ctoon.cpp
/*
 * CToon C++ benchmark.
 *
 * We're the reference implementation for the C++ binding, so this isn't run
 * against any competing library — it just tracks the wrapper's own
 * throughput (and the RAII/exception overhead over the raw C API) across
 * changes to ctoon.hpp.
 *
 * Same methodology as every other language benchmark in this repo — see
 * benchmarks/README.md — so the numbers line up with the C, Python, and Go
 * results for the same corpus.
 */

#define CTOON_ENABLE_JSON 1
#include "ctoon.hpp"

#include <cstdio>
#include <ctime>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>

#ifndef CTOON_BENCH_MANIFEST
#error "CTOON_BENCH_MANIFEST must be defined at compile time"
#endif

#ifndef CTOON_BENCH_REPEATS
#define CTOON_BENCH_REPEATS 20
#endif

namespace {

struct BenchFile {
    std::string path;
    std::string json;
    std::string toon; // filled in during the pre-pass
};

double now_seconds() {
    return static_cast<double>(std::clock()) / static_cast<double>(CLOCKS_PER_SEC);
}

bool read_whole_file(const std::string &path, std::string &out) {
    std::ifstream f(path, std::ios::binary);
    if (!f) return false;
    std::ostringstream ss;
    ss << f.rdbuf();
    out = ss.str();
    return true;
}

} // namespace

int main() {
    std::ifstream manifest(CTOON_BENCH_MANIFEST);
    if (!manifest) {
        std::fprintf(stderr, "Cannot open manifest: %s\n", CTOON_BENCH_MANIFEST);
        return 1;
    }

    std::vector<BenchFile> files;
    std::string line;
    while (std::getline(manifest, line)) {
        if (line.empty()) continue;
        BenchFile bf;
        bf.path = line;
        if (!read_whole_file(line, bf.json)) continue;
        files.push_back(std::move(bf));
    }

    if (files.empty()) {
        std::fprintf(stderr, "Corpus manifest is empty - nothing to benchmark.\n");
        return 1;
    }

    std::size_t total_json_bytes = 0;
    for (const auto &f : files) total_json_bytes += f.json.size();

    std::printf("CToon C++ Benchmark\n");
    std::printf("Corpus: %zu files, %.2f MB (JSON)\n\n",
                 files.size(), static_cast<double>(total_json_bytes) / 1e6);

    // ── Untimed pre-pass: produce TOON text for each file ──
    std::size_t total_toon_bytes = 0;
    for (auto &f : files) {
        try {
            auto doc = ctoon::document::from_json(f.json);
            f.toon = doc.to_string().str();
            total_toon_bytes += f.toon.size();
        } catch (const ctoon::error &) {
            // Skip files that don't round-trip (e.g. a scalar-only JSON
            // document, which some corpora include as edge cases).
        }
    }

    // ── Phase A: JSON -> TOON ──
    double t0 = now_seconds();
    long ops_a = 0;
    for (int rep = 0; rep < CTOON_BENCH_REPEATS; rep++) {
        for (const auto &f : files) {
            try {
                auto doc = ctoon::document::from_json(f.json);
                auto result = doc.to_string();
                (void)result.size();
                ops_a++;
            } catch (const ctoon::error &) {
                // counted as skipped, not an error - matches the pre-pass
            }
        }
    }
    double t_json_to_toon = now_seconds() - t0;
    double bytes_a = static_cast<double>(total_json_bytes) * CTOON_BENCH_REPEATS;

    // ── Phase B: TOON -> JSON (using the pre-pass output) ──
    t0 = now_seconds();
    long ops_b = 0;
    for (int rep = 0; rep < CTOON_BENCH_REPEATS; rep++) {
        for (const auto &f : files) {
            if (f.toon.empty()) continue;
            try {
                auto doc = ctoon::document::parse(f.toon);
                auto result = doc.to_json(2);
                (void)result.size();
                ops_b++;
            } catch (const ctoon::error &) {
            }
        }
    }
    double t_toon_to_json = now_seconds() - t0;
    double bytes_b = static_cast<double>(total_toon_bytes) * CTOON_BENCH_REPEATS;

    // ── Report ──
    std::printf("%-14s %12s %14s %12s\n", "Operation", "Throughput", "Docs/sec", "Total time");
    std::printf("%-14s %9.2f MB/s %14.0f %9.4f s  (x%d reps)\n",
                 "JSON -> TOON", bytes_a / t_json_to_toon / 1e6,
                 static_cast<double>(ops_a) / t_json_to_toon, t_json_to_toon, CTOON_BENCH_REPEATS);
    std::printf("%-14s %9.2f MB/s %14.0f %9.4f s  (x%d reps)\n",
                 "TOON -> JSON", bytes_b / t_toon_to_json / 1e6,
                 static_cast<double>(ops_b) / t_toon_to_json, t_toon_to_json, CTOON_BENCH_REPEATS);

    return 0;
}
